### PR TITLE
qsort

### PR DESCRIPTION
**Pull request by janisozaur**
Expect near-instant startup times now. @Gymnasiast this one is for you and your 1.4GB of objects.

### CODE REVIEW

File: src/object_list.c [hunk 1]
```diff
@@ -29,6 +29,9 @@
 #include "world/scenery.h"
 #include "world/water.h"
 
+// for qsort()
+#include <stdlib.h>
```
Comment (IntelOrca) [inline src/object_list.c]: Already included in rct2.h

Comment (janisozaur) [inline src/object_list.c]: I was initially testing with Linux's `qsort_r` and compiler complained about the function being undefined. Updated now.

File: src/object_list.c [hunk 2]
```diff
@@ -134,29 +147,29 @@ static void object_list_sort()
 		destFilter = newFilters;
 	}
 
+	_entryCache = malloc(numObjects * sizeof(rct_object_entry*));
+	size_t *sortLUT = malloc(numObjects * sizeof(size_t));
```
Comment (IntelOrca) [inline src/object_list.c]: What does LUT stand for?

Comment (janisozaur) [inline src/object_list.c]: Look-Up Table

File: src/object_list.c [hunk 3]
```diff
@@ -108,12 +111,22 @@ static void get_plugin_path(utf8 *outPath)
 	strcat(outPath, "plugin.dat");
 }
 
+rct_object_entry **_entryCache = NULL;
```
Comment (IntelOrca) [inline src/object_list.c]: `static`

Comment (janisozaur) [inline src/object_list.c]: done

File: src/object_list.c [hunk 4]
```diff
@@ -134,37 +144,29 @@ static void object_list_sort()
...
-				}
-			}
-		}
-		entrySize = object_get_length(lowestEntry);
-		memcpy(destEntry, lowestEntry, entrySize);
+		entrySize = object_get_length(_entryCache[sortLUT[i]]);
+		memcpy(destEntry, _entryCache[sortLUT[i]], entrySize);
 		destEntry = (rct_object_entry*)((int)destEntry + entrySize);
 		if (_installedObjectFilters)
 			destFilter[i] = _installedObjectFilters[lowestIndex];
```
Comment (janisozaur) [inline src/object_list.c]: `lowestIndex` is leftover from previous version, I'll fix it in a moment

**Comment by IntelOrca** [comment]
If you wanted to avoid global variables for the sort function to use, you could instead sort a list of (new struct tuple (object_entry, original ID))

**Comment by janisozaur** [comment]
I fixed the `lowestIndex` issue.

It's fine, I can make do with this local variable.

**Comment by janisozaur** [comment]
If it weren't for `destFilter` which needs the index, I could simply sort `entryCache` (note: local, not static-global) instead of `sortLUT`, which would be the best solution, but it doesn't seem possible.

I'm happy with it now, even if your suggestion may be sound. It's near-instant now, whereas it took some non-negligible time before.

**Comment by Gymnasiast** [comment]
Building plugin.dat with current version: 13 seconds.
Building plugin.dat with your qsort: 3 seconds. :+1: 
I think the people at NEdesigns are going to love it as well. :)

**Comment by janisozaur** [comment]
3 seconds? That's ages, should be instant! :P

**Comment by IntelOrca** [comment]
@janisozaur well there's a challenge for you, make `object_list_load` instant

**Pull request merged**